Consideration for Marketplace Shops on Different URLs

pmn4 opened the pull request:
This PR is to address Merchant Shops which live on a different URL than the Primary Shop.
For example, my Primary Shop is hosted at https://reserve.brewline.io and a Merchant Shop at https://twolights.brewline.io. Visiting the latter shows the branding and products of the Primary Shop.
Opening developer tools on the Merchant Shop and running `ReactionCore.getPrimaryShopId() === ReactionCore.getShopId()` yields `true` (the merchant shop has its `domains` attribute set properly).

Digging into the code yielded that `Meteor.absoluteUrl` relies on `$ROOT_URL` which is a problem when using the domain as an identifier. To solve this, I capture the domain/hostname during `onConnection` (server-side) and via `window.location.hostname` (client-side).

I then wrapped `Meteor.absoluteUrl` in a function called `Reaction.absoluteUrl`, passing a `rootUrl` option, and doing a repo-wide find/replace.

Lastly, on `Meteor.startup` (client side), I subscribe to `"Shop"` (previously, `"PrimaryShop"`) to ensure the shop is ready before rendering.

---
**_disclaimer:_** I learned a ton doing this, but there are definitely still gaps in my understanding of Meteor and Reaction. Feel free to highlight anything that may look fishy!

**_further disclaimer:_** I don't know all of the Reaction features yet, so there is a lot of regression testing to be done before merging.

lastly, where should I write unit tests for these new methods?
thanks for the consideration!

spencern commented:
@aaronjudd @ticean are you still planning to review this one?

pmn4 commented on client/modules/core/main.js:
```diff
@@ -158,6 +159,46 @@ export default {
+   * @return {String} Domain/hostname for the given options
+   */
+  getDomain(options) {
+    return this.absoluteUrl("", options).split("/")[2].split(":")[0];
```
This is copied it from [another part of the code-base](https://github.com/reactioncommerce/reaction/blob/b297619593a15dc0fd751c343970cd53ae89c42b/client/modules/core/main.js#L97) to keep a single technique for extracting the domain.

pmn4 replied:
for sure. I didn't want to create a second pattern for doing the same thing, so I copy/pasted this logic.
I will make this `url.parse` and attempt to update the client-side test as well, assuming the URL library is available there.

pmn4 commented on imports/plugins/core/router/client/startup.js:
```diff
@@ -14,22 +14,21 @@ Meteor.startup(function () {
   // or missing shop data throughout the app.
-  // TODO: Revisit subscriptions manager usage and waiting for shops to exist client side before rendering.
-  const primaryShopSub = Meteor.subscribe("PrimaryShop");
+  const shopSub = Meteor.subscribe("Shop");
```
looking closer, I don't believe this change is necessary (although, from what I can tell, not harmful either)

spencern replied:
If this change isn't necessary, let's remove it. Not sure what it might affect in 3rd party plugins or other shops since it's a managed subscription.

spencern commented on client/modules/core/main.js:
```diff
@@ -158,6 +159,46 @@ export default {
+   * @return {String} URL for the given path and options
+   */
+  absoluteUrl(path, options) {
+    const o = _.extend({}, options);
```
Can use `Object.assign({}, options)` instead of lodash `extend` here to clone as it doesn't look like there are any deep properties being mutated.

Also would prefer a more explicit variable name such as opts.

spencern commented on client/modules/core/main.js:
```diff
@@ -158,6 +159,46 @@ export default {
+    const o = _.extend({}, options);
+    const domain = this._shopDomain.get();
+
+    if (!("rootUrl" in o) && domain) {
```
Perhaps more preference, but I'd prefer if we abstracted this `!("rootUrl" in o)` logic to it's own variable to make this line more readable.
```
const hasRootUrl = "rootUrl" in o;

if (domain && !hasRootUrl) {
  // do stuff
```

spencern commented on server/api/core/core.js:
```diff
@@ -67,6 +74,33 @@ export default {
+   * @param {String} options.rootUrl Override the default ROOT_URL from the server environment. For example: "`http://foo.example.com`"
+   */
+  absoluteUrl(path, options) {
+    const o = _.extend({}, options);
```
Comments on client method apply here too.

spencern commented on client/modules/core/main.js:
```diff
@@ -158,6 +159,46 @@ export default {
+   * @param {String} options.rootUrl Override the default ROOT_URL from the server environment. For example: "`http://foo.example.com`"
+   * @return {String} URL for the given path and options
+   */
+  absoluteUrl(path, options) {
```
Generally, this is the right approach, wrapping the Meteor method in a Reaction method so that if we migrate away from Meteor, these APIs don't change.

spencern commented on server/publications/collections/shops.js:
```diff
@@ -12,6 +12,15 @@ Meteor.publish("PrimaryShop", function () {
   });
 });
 
+Meteor.publish("Shop", function () {
```
Curious to understand what the change here is intended to do and why the old publication was left alone even though it's subscription was removed.

pmn4 replied:
oops! just an oversight

spencern commented:
@aaronjudd @ticean did you guys still want to take a look at this PR?

pmn4 commented:
@spencern I had to do some port forwarding magic to simulate this locally on my mac.

first, edit `/etc/hosts`
```
40.0.0.1 local.reactioncommerce.com
```
then, run
``` bash
sudo ifconfig lo0 40.0.0.1 alias
echo "rdr pass on lo0 inet proto tcp from any to 40.0.0.1 port 80 -> 127.0.0.1 port 3000" | sudo pfctl -ef -
```
then, set up two shops, one with `domains: ["localhost"]` and the other with `domains: ["local.reactioncommerce.com", "localhost"]` (localhost is optional, but without it, you cannot switch to it from the Primary Shop)

aaronjudd commented:
@spencern I'm good with the direction, and don't need to be a blocker.

pmn4 commented:
@spencern I have updated this PR with a few things that I learned from doing other work, namely, instead of setting a `ReactiveVar` onConnection -- I'm not positive `ReactiveVar`s work the way I was expecting them to on the server -- I went to direct access of the DDP Connection.

Next, I removed the unnecessary changes and made the edits you suggested.

Focus your attention on the [server-side code](https://github.com/reactioncommerce/reaction/pull/3332/files#diff-8efd71b38bf2153707a7eb6da1e58579) as nearly everything else is either:

1. find/replace of `Meteor.absoluteUrl` with `Reaction.absoluteUrl`
2. code you have already reviewed
3. tests

aaronjudd commented:
@reactioncommerce/community can we move this forward for 1.9? Shepherds needed...

zenweasel commented:
@pmn4 Sorry for the delay here Pat. If we can get these conflicts resolved I will make sure we get this moving forward. Thanks for your patience.

zenweasel commented:
I can't resolve conflicts here since it's on a remote repository to which I don't have write rights

zenweasel commented:
@pmn4 This one is on your remote repository so we can't resolve conflicts. If you could resolve these we can try to get these moved forward quickly. Thanks for your patience with these.

zenweasel commented:
@pmn4 We have conflicts again here Pat. Can we get them resolved and try to get this tested and merged.

zenweasel commented:
@pmn4 Also we really need to get some test steps so that we have an easy way to confirm that this code functions as expected. Also test steps around exactly we should be testing for regression testing. This is going to be the main hurdle to getting this merged. Thanks so much.

pmn4 commented on client/modules/core/__tests__/main.js:
```diff
@@ -0,0 +1,90 @@
+import { expect } from "meteor/practicalmeteor:chai";
```
I am unable to even run this test at the moment. `npm test` is not playing nicely with babel, or something. I have spent many hours trying to get these tests to run with no luck. I would appreciate all suggestions.
to repro, just run `npm test`

pmn4 commented:
To test the primary use case, add an entry to your `/etc/hosts` file
```
40.0.0.1 local.reactioncommerce.com
```
then do some port forwarding (h/t google.com for this one)
```
sudo ifconfig lo0 40.0.0.1 alias
echo "rdr pass on lo0 inet proto tcp from any to 40.0.0.1 port 80 -> 127.0.0.1 port 3000" | sudo pfctl -ef -
```
create a marketplace shop and _add_ the domain `local.reactioncommerce.com` (you'll need to do this directly in the database). Give it a brand image and a different name to illustrate the point.
on the master branch, visit `local.reactioncommerce.com`. Notice that the Brand Image/Name in the header is that of the Primary Shop, not the Marketplace Shop.
on this branch, you should get the Marketplace Shop by visiting that URL.

zenweasel commented:
@pmn4 There are conflicts here. Ping me after you resolve these so we can try to get this test/approved.

pmn4 commented:
hey @zenweasel, sorry for the delay. haven't had much time lately to work on these.
Are you able to run the `jest` test suite? I added a few tests there, but had to heavily modify the package.json and even some unrelated source code to get it working. If we are unable to run the client-side tests, I will remove that test file

zenweasel commented:
@pmn4 So the jest tests are run in CI now (`npm run test:unit`) and it looks like they are failing?

pmn4 commented:
I was able to get the tests running locally, however to do so, I had to comment out code which requires `lib/collections`. The suite was failing on a lint or perhaps a transpile error: 
```
SyntaxError: reaction/lib/collections/collections.js: 'import' and 'export' may only appear at the top level (169:2)
```
I have spent several hours trying to figure this out. any suggestions would be GREATLY appreciated.

pmn4 commented on client/modules/core/main.js:
```diff
@@ -164,6 +164,57 @@ export default {
+    if (!options && typeof path === 'object') {
+      options = path;
+      path = undefined;
+    }
```
this block of code mirrors `Meteor.absoluteUrl` to ensure their method signatures are the same

pmn4 commented on imports/test-utils/__mocks__/meteor/alanning:roles.js:
```diff
@@ -0,0 +1,13 @@
+export const Roles = {
```
created the following files to get jest tests passing.
we tell jest to replace all `meteor/` modules with local mocks, and some were missing.

pmn4 commented on client/modules/core/__tests__/main.js:
```diff
@@ -0,0 +1,83 @@
+import { Meteor } from "meteor/meteor";
```
these tests are likely to fail on circle:ci. to get them running locally, I had to comment out the following lines in `main.js`:
``` JavaScript
 // import { localeDep } from "/client/modules/i18n";
 // import { Packages, Shops, Accounts } from "/lib/collections";
 // import { Router } from "/client/modules/router";
```

spencern commented:
@aldeed would you mind taking a look at this at some point to see if you can figure out why @pmn4's unit tests are failing?

aldeed commented on client/modules/core/__tests__/main.js:
```diff
@@ -0,0 +1,83 @@
+
+import main from "../main";
+
+describe("Client/API/Core", () => {
```
Change all `it` to `test`.

It also would be better to move `shopDomain`, `getDomain`, and `absoluteUrl` into their own file (maybe `domain.js`. You can export an object with just those three props and in `main.js`, import it and spread it onto the main object. This will help isolate better and keep dependencies to a minimum, which should help the non-Meteor unit tests pass more easily.

We also use the `.test.js` convention rather than `__tests__` folder. So you should have:

```
domain.js
domain.test.js
```

pmn4 replied:
sounds good

pmn4 commented on client/modules/core/main.js:
```diff
@@ -98,9 +100,7 @@ export default {
-        // get domain (e.g localhost) from absolute url (e.g http://localhost:3000/)
-        const [, , host] = Meteor.absoluteUrl().split("/");
-        const [domain] = host.split(":");
+        this._shopDomain.set(document.location.hostname); // client side code... is this ok??
```
nice. will do.

pmn4 commented on imports/plugins/core/files/server/fileCollections.js:
```diff
@@ -24,7 +24,7 @@ async function lazyLoadSharp() {
 
 FileRecord.downloadEndpointPrefix = "/assets/files";
-FileRecord.absoluteUrlPrefix = Meteor.absoluteUrl();
+FileRecord.absoluteUrlPrefix = Meteor.absoluteUrl(); // TODO: Reaction.absoluteUrl()
```
this line was added after the PR was created and I haven't had time to test whether the change to `Reaction.absoluteUrl` would work in this context. I'm confident in the change at runtime but haven't investigated initialization time yet.

pmn4 replied:
@aldeed maybe you can help me here. adding an import statement results in an undefined `Reaction`, which I presume means this code runs before we build `Reaction`. Should it? I'm curious whether this code should live inside a function, and kicked off in a startup script once all code has been loaded.

aldeed replied:
@pmn4 `Reaction` obj should never be undefined when imported. Are you sure the import was correct? `import { Reaction } from "/server/api";`

pmn4 replied:
@aldeed I haven't dug in yet, but trying to start the app yields:
``` 
W20180425-13:38:34.514(-4)? (STDERR) TypeError: Cannot read property 'absoluteUrl' of undefined
W20180425-13:38:34.514(-4)? (STDERR)     at fileCollections.js (imports/plugins/core/files/server/fileCollections.js:27:41)
W20180425-13:38:34.514(-4)? (STDERR)     at fileEvaluate (packages/modules-runtime.js:343:9)
W20180425-13:38:34.514(-4)? (STDERR)     at require (packages/modules-runtime.js:238:16)
W20180425-13:38:34.515(-4)? (STDERR)     at index.js (imports/plugins/core/files/server/index.js:1:57)
W20180425-13:38:34.515(-4)? (STDERR)     at fileEvaluate (packages/modules-runtime.js:343:9)
W20180425-13:38:34.515(-4)? (STDERR)     at require (packages/modules-runtime.js:238:16)
W20180425-13:38:34.515(-4)? (STDERR)     at email.js (server/api/core/email/email.js:1:567)
W20180425-13:38:34.516(-4)? (STDERR)     at fileEvaluate (packages/modules-runtime.js:343:9)
W20180425-13:38:34.516(-4)? (STDERR)     at require (packages/modules-runtime.js:238:16)
W20180425-13:38:34.516(-4)? (STDERR)     at index.js (server/api/core/email/index.js:1:78)
W20180425-13:38:34.516(-4)? (STDERR)     at fileEvaluate (packages/modules-runtime.js:343:9)
W20180425-13:38:34.517(-4)? (STDERR)     at require (packages/modules-runtime.js:238:16)
W20180425-13:38:34.517(-4)? (STDERR)     at index.js (server/api/core/index.js:1:334)
W20180425-13:38:34.517(-4)? (STDERR)     at fileEvaluate (packages/modules-runtime.js:343:9)
W20180425-13:38:34.517(-4)? (STDERR)     at require (packages/modules-runtime.js:238:16)
W20180425-13:38:34.518(-4)? (STDERR)     at index.js (server/api/index.js:1:190)
W20180425-13:38:34.518(-4)? (STDERR)     at fileEvaluate (packages/modules-runtime.js:343:9)
W20180425-13:38:34.518(-4)? (STDERR)     at require (packages/modules-runtime.js:238:16)
W20180425-13:38:34.518(-4)? (STDERR)     at core.js (lib/api/core.js:10:10)
W20180425-13:38:34.519(-4)? (STDERR)     at fileEvaluate (packages/modules-runtime.js:343:9)
W20180425-13:38:34.519(-4)? (STDERR)     at require (packages/modules-runtime.js:238:16)
W20180425-13:38:34.519(-4)? (STDERR)     at index.js (lib/api/index.js:1:313)
```

pmn4 replied:
there is a line already which imports `Logger` from "/server/api". I added Reaction to that statement:
``` JavaScript
import { Logger, Reaction } from "/server/api";
console.log("Immediate", !!Logger, !!Reaction);
setTimeout(() => console.log("Later", !!Logger, !!Reaction), 5000);
```
the console log yields
```
Immediate false false
// ... some other console messages ...
Later true true
```

This behavior is consistent with a circular dependency.
I suspect it was never identified previously because the depencies are eventually resolved. This is likely one of the few places where we import code and try to use it immediately.

aldeed replied:
@pmn4 Are you able to identify where the circular dependency is and unravel it? Or for example could you import your new domains object directly rather than the whole `Reaction` object?

pmn4 commented on client/modules/core/main.js:
```diff
@@ -164,6 +164,64 @@ export default {
+   * getDomain
+   * @summary extracts the domain name from the absoluteUrl.
+   * @param {Object} [options] inheritied from absoluteUrl
+   * @param {Boolean} options.secure Create an HTTPS URL.
```
good point!

zenweasel commented:
@pmn4 Is this ready for another look/test?

pmn4 commented on imports/plugins/core/files/server/fileCollections.js:
```diff
@@ -15,6 +15,16 @@ import {
+// import { Reaction } from "/server/api";
+import { DomainsMixin as ReactionDomainMethods } from "/server/api/domains";
+
+FileRecord.downloadEndpointPrefix = "/assets/files";
```
I moved these line up to keep them close to the import and comment.

pmn4 commented:
@zenweasel I have not had time to investigate the exact cause of the circular dependency, but I did update the code as @aldeed suggested.

zenweasel commented:
@aldeed @spencern I think this PR is good and important for marketplace and it's been open for 8 months now. I wonder if we shouldn't take ownership of it and try to get it merged? Willing to do whatever it takes in terms of testing or whatever.

Let me know what you think

pmn4 commented:
thanks @zenweasel! I've done my best to keep this mergeable. lmk what I can do to help or if you have any questions

spencern commented:
Agreed @zenweasel @pmn4 - apologies that we've not been able to get this merged yet.

spencern commented:
To echo what @zenweasel said, this is a PR I'd like to see merged. Unfortunately because it's on your fork, we can't really help it over the line by resolving conflicts etc.

pmn4 commented:
looking forward to it. I'll work on the conflicts today

pmn4 commented on imports/plugins/core/files/server/fileCollections.js:
```diff
@@ -10,6 +10,7 @@ import {
   RemoteUrlWorker,
   TempFileStoreWorker
 } from "@reactioncommerce/file-collections";
+import { AbsoluteUrlMixin } from "/imports/plugins/core/core/server/Reaction/absoluteUrl";
```
importing the mixin directly to avoid a circular dependency.
I have found that circular dependencies are _eventually_ resolved once both files have been fully loaded. We rarely notice them (if they exist) because most calls are made in some initialization step once all of the code is loaded. this, on the other hand, is being called immediately, before the `Reaction` API has not yet been fully loaded.

pmn4 commented:
@zenweasel @spencern y'all shuffled around a bunch of files on me!
Waiting for tests to run, but this branch is now conflict-free of release-1.14.0

zenweasel commented:
@pmn4 I still can't fast-forward so I can't pull this locally to test. Can you update your branch to `release-1.14.0` again?

pmn4 commented:
ugh!
off to bed now. will update again in the morning (EST)

pmn4 commented:
that's strange... why does GH think it is mergeable?

zenweasel commented:
Because there's no conflicts. Which is different from being able to FF, which means that only these commits must be between your branch and `release-1.14.0`.

```
 »» git checkout -b pmn4-pmn4/Reaction-absoluteUrl release-1.14.0
 »» git pull https://github.com/pmn4/reaction.git pmn4/Reaction-absoluteUrl
remote: Counting objects: 257, done.
remote: Compressing objects: 100% (45/45), done.
remote: Total 257 (delta 128), reused 174 (delta 128), pack-reused 80
Receiving objects: 100% (257/257), 91.67 KiB | 210.00 KiB/s, done.
Resolving deltas: 100% (142/142), completed with 54 local objects.
From https://github.com/pmn4/reaction
 * branch                pmn4/Reaction-absoluteUrl -> FETCH_HEAD
fatal: Not possible to fast-forward, aborting.
```

zenweasel commented:
I'm guessing that every time we merge something to `release-1.14.0` that means I can't fast-forward anymore

pmn4 commented:
@aldeed gone. I did little more than copy the Meteor code (and appease eslint). I struggled a bit with the file location. lmk if you think there is a better location for it.

zenweasel commented:
@aldeed Can you take a look at this and dismiss your review if this is ready to go?

spencern commented:
We're so close, some new conflicts with our removed plugins I think.

spencern commented:
Thanks @pmn4, merging 🎉
